src/main/java/com/linkedin/kmf/consumer/NewConsumer.java

```diff
@@ -15,33 +15,94 @@
 import java.util.Iterator;
 import java.util.Map;
 import java.util.Properties;
+import java.util.concurrent.ExecutionException;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
 import org.apache.kafka.clients.consumer.ConsumerRecord;
 import org.apache.kafka.clients.consumer.KafkaConsumer;
 import org.apache.kafka.clients.consumer.OffsetAndMetadata;
 import org.apache.kafka.clients.consumer.OffsetCommitCallback;
 import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.internals.Topic;
+import org.apache.kafka.common.utils.Utils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-/*
- * Wrap around the new consumer from Apache Kafka and implement the #KMBaseConsumer interface
+
+/**
+ * Wraps around the new consumer from Apache Kafka and implements the #KMBaseConsumer interface
  */
 public class NewConsumer implements KMBaseConsumer {
 
   private final KafkaConsumer<String, String> _consumer;
   private Iterator<ConsumerRecord<String, String>> _recordIter;
-  private static final Logger LOG = LoggerFactory.getLogger(NewConsumer.class);
+  private static final Logger LOGGER = LoggerFactory.getLogger(NewConsumer.class);
   private static long lastCommitted;
 
-  public NewConsumer(String topic, Properties consumerProperties) {
+  // package private permission for NewConsumerTest to access.
+  String _consumerGroupPrefix = "__shadow_consumer_group-";
+  int _consumerGroupSuffixCandidate = 0;
+  String _targetConsumerGroupId;
```

Comment:
There's no reason for these to be instance variables.  These are never referred to after being used once.   PartitionsFor() doesn't need to be an instance member.   These choices make this functionality hard to test.

String _consumerGroupPrefix = "__shadow_consumer_group-";,  This never changes, it can be a constant. If it needs to be referred to for testing purposes then mark it as "static final".  If it's package private then tests can access it as well.

```java
class SomeUtilClass {
/** documentation */
static int partitionFor(String groupId, int consumerOffsetsTopicPartitions) {

    LOGGER.debug("Hashed and modulo output: {}", groupId.hashCode());
    return Utils.abs(groupId.hashCode()) % consumerOffsetsTopicPartitions;
  }

/** documentation */
  // Instead of making targetGroupId an instance variable and then assigning it some value which this then looks up
  // it can just be a parameter to a method
  static String findCollision(String targetGroupId, AdminClient adminClient)
      throws ExecutionException, InterruptedException {
    if (targetGroupId.equals("")) {
      throw new IllegalArgumentException("The target consumer group identifier cannot be empty: " + targetGroupId);
    }

 int numOffsetsTopicPartitions = adminClient.describeTopics(Collections.singleton(Topic.GROUP_METADATA_TOPIC_NAME))
	
        .values()
        .get(Topic.GROUP_METADATA_TOPIC_NAME)
        .get()
        .partitions()
        .size();

   // Extract invariant from loop
   int targetConsumerOffsetsPartition = partitionsFor(targetGroupId, numOffsetsTopicPartitions);

   //  This doesn't need to be an instance variable because we throw this out this value at the end of computation
    int groupSuffix = 0;

    // Extract return value so it's not computed twice, this reduces the possibility of bugs
    String newGroupName = null;

    // Use while(true) otherwise halting condition is hard to read.
    while (true) {
      //TODO:  could play fancy StringBuilder games here to make this generate less garbage
       newGroupName = GROUP_SUFFIX_CANDIDATE + Integer.toString(groupSuffix++);
       int newGroupNamePartition = partitionsFor(newGroupName, numOffsetsTopicPartitions);
        if (targetConsumerOffsetsPartition == newGroupNamePartition) {
           break;
        }
    }
   return newGroupName;
}

```

I ended up writing this code as a comment.  It probably doesn't compile.  It's here to convey ideas.
For testing I would just run partitionsFor() on the result to make sure they are the same and that they groupIds are not the same.

Reply:
`       newGroupName = GROUP_SUFFIX_CANDIDATE + Integer.toString(groupSuffix++);` 
--> did you mean to put `GROUP_PREFIX_CANDIDATE`?

Reply:
Other than that, I'm content with adopting the above class that's proposed.